From a112113e68c1bed72d13fc528fb429837de4f12e Mon Sep 17 00:00:00 2001
From: Your Name <you@example.com>
Date: Thu, 19 Jan 2017 13:46:16 +0000
Subject: [PATCH] Changed parameters to values for maximum precision

---
 CMakeLists.txt | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 64ed3b7..9ae9d48 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -61,7 +61,7 @@ set(VERSION_PATCH 1)
 
 # TLSH uses only half the counting buckets.
 # It can use all the buckets now.
-set(TLSH_BUCKETS_128 1)
+set(TLSH_BUCKETS_256 1)
 if(TLSH_BUCKETS_48 EQUAL 1)
   set(TLSH_HASH "min hash")
   add_definitions(-DBUCKETS_48)
@@ -77,7 +77,7 @@ endif()
 
 # TLSH uses 1 byte checksum. The collision rate is 1 in 24.
 # It can use 3 bytes checksum now. That collision rate in 1 in 5800.
-set(TLSH_CHECKSUM_1B 1)
+set(TLSH_CHECKSUM_1B 3)
 if(TLSH_CHECKSUM_1B EQUAL 1)
   set(TLSH_CHECKSUM "1 byte checksum")
 else()
-- 
2.7.4

